apps/src/music/README.md

```diff
@@ -31,30 +32,71 @@ This is the client-side code for **Music Lab**, as instantiated by a `music` lev
+We built keyboard shortcuts to cycle through the various possibilities, and it proved invaluable. After using the app for a while and playing with these variants, we settled on the options that felt the best. It was a great way to solve the open questions through regular usage, and saved on abstract debates.
 
 ## URL parameter menu
 
 We added a variety of internal URL parameters to also assist in development, and built a small page at https://studio.code.org/musiclab/menu to generate the correct URL.
+
+## Compilation, Execution, and Playback
+
+When a user edits their code or runs the project, Blockly blocks are compiled to JavaScript and, if any changes are detected, immediately executed. This is handled by the [MusicBlocklyWorkspace](./blockly/MusicBlocklyWorkspace.ts) component.
+
+During execution, the execution runtime is provided a [Sequencer](./player/sequencer/Sequencer.ts) whose functions are called by the generator functions of each block ([example](./blockly/blocks/simple2.js#L90)). The Sequencer encapsulates all sequencing logic and there are a few different implementations for the different programming models supported (the [Simple2Sequencer](./player/sequencer/Simple2Sequencer.ts) being the primary one, corresponding to the "simple2" model). As a result of these function calls, the Sequencer assembles a list of [PlaybackEvents](./player/interfaces/PlaybackEvent.ts) based on the contents of each block, representing the full set of events in the song.
+
+When playback starts, the list of PlaybackEvents is handed to the [MusicPlayer](./player/MusicPlayer.ts) which handles all playback related functions. Internally, the MusicPlayer delegates to either a [ToneJSPlayer](./player/ToneJSPlayer.ts) which uses the ToneJS library, or a [SamplePlayer](./player/SamplePlayer.ts) which uses Web Audio APIs directly. The SamplePlayer code was used initially, while the ToneJSPlayer code was added more recently when more advanced functionality was required (such as pitch shifting and time stretching). Both player implementations are abstracted behind an [AudioPlayer](./player/types.ts) interface.
+
+### Handling Updates Mid-Playback
+
+Music Lab supports live updates during playback. When code changes are detected, code is recompiled and executed in the same manner as above. The new list of PlaybackEvents is handed to the player, along with a flag that indicates to the player that it should replace its current set of events with this new set. The player cancels all pending events, and requeues the new set of events. Depending on the specific player implementation, the user hears the updates almost immediately, or latest when the next sound starts.
```

Comment:
> or latest when the next sound starts

later?

Reply:
Yeah this is worded weirdly, I'll just remove - meaning to say that with the ToneJSPlayer, updates are heard immediately and with the SamplePlayer, updates are heard when the next sounds starts.